File: python/src/truss/budget/memory_store.py
from __future__ import annotations

import threading
from truss.budget.config import BudgetWindow
from truss.budget.ledger import LedgerEntry, UsageReport


class InMemoryStore:
    def __init__(self) -> None:
        self._entries: list[LedgerEntry] = []
        self._lock = threading.RLock()

    def record(self, entry: LedgerEntry) -> None:
        with self._lock:
            self._entries.append(entry)

    def usage(self, key: str, window: BudgetWindow) -> UsageReport:
        with self._lock:
            matching = [
                e for e in self._entries
                if e.session_id == key or e.user_id == key or e.agent_name == key
            ]
            total_tokens = sum(e.total_tokens for e in matching)
            total_cost = sum(e.cost_usd for e in matching)
        return UsageReport(
            key=key,
            total_tokens=total_tokens,
            total_cost_usd=total_cost,
            window=window,
        )

    def flush(self) -> None:
        pass

File: typescript/src/budget/memory-store.ts
import { LedgerEntry, LedgerStore, UsageReport } from './ledger.js';
import { BudgetWindow } from './config.js';

export class InMemoryStore implements LedgerStore {
  private entries: LedgerEntry[] = [];

  record(entry: LedgerEntry): void {
    this.entries.push(entry);
  }

  usage(key: string, window: BudgetWindow): UsageReport {
    const matching = this.entries.filter(
      e => e.sessionId === key || e.userId === key || e.agentName === key,
    );
    const totalTokens = matching.reduce((s, e) => s + e.inputTokens + e.outputTokens, 0);
    const totalCostUsd = matching.reduce((s, e) => s + e.costUsd, 0);
    return { key, totalTokens, totalCostUsd, window, pctUsed: 0 };
  }

  flush(): void {}
}
